### PR TITLE
Add function rte_pmd_bnxt_set_tx_loopback and rte_pmd_bnxt_set_all_queues_drop_en

### PR DESCRIPTION
**Pull request by steevenlee**
Moved code review from Stephen's personal branch to here to involve team.

### CODE REVIEW

File: drivers/net/bnxt/bnxt_ethdev.c [hunk 1]
```diff
@@ -819,6 +820,78 @@ static int bnxt_rss_hash_update_op(struct rte_eth_dev *eth_dev,
 	return 0;
 }
 
+int rte_pmd_bnxt_set_tx_loopback(uint8_t port, uint8_t on)
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt_ethdev.c]: Stephen raised this question in another review, this method invoked by VFd, are we going to support it in HWRM or just delete it?

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 2]
```diff
@@ -877,6 +881,38 @@ int bnxt_hwrm_vnic_cfg(struct bnxt *bp, struct bnxt_vnic_info *vnic)
...
+	vnic->fw_rss_cos_lb_ctx = rte_le_to_cpu_16(resp->rss_rule);
+	// vnic->cos_rule = rte_le_to_cpu_16(resp->cos_rule);
+	// vnic->lb_rule = rte_le_to_cpu_16(resp->lb_rule);
+	vnic->mru = rte_le_to_cpu_16(resp->mru);
+	vnic->func_default = rte_le_to_cpu_32(
+			resp->flags) & HWRM_VNIC_QCFG_OUTPUT_FLAGS_DEFAULT;
+	vnic->vlan_strip = rte_le_to_cpu_32(
+			resp->flags) & HWRM_VNIC_QCFG_OUTPUT_FLAGS_VLAN_STRIP_MODE;
+	vnic->bd_stall = rte_le_to_cpu_32(
+			resp->flags) & HWRM_VNIC_QCFG_OUTPUT_FLAGS_BD_STALL_MODE;
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt_hwrm.c:923]: although bd_stall can be set here and read out correctly later, traffic still pass despite of bd_stall flag, not sure how to debug chimp under dpdk or how to involve chimp team on this...

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 3]
```diff
@@ -960,6 +996,56 @@ int bnxt_hwrm_vnic_rss_cfg(struct bnxt *bp,
...
+
+	num_vnic_ids = rte_le_to_cpu_32(resp->vnic_id_cnt);
+
+	/* Retrieve VNIC, update bd_stall then update */
+
+	for (i = 0; i < num_vnic_ids; i++) {
+		memset(&vnic, 0, sizeof(struct bnxt_vnic_info));
+		vnic.fw_vnic_id = rte_le_to_cpu_16(vnic_ids[i]);
+		rc = bnxt_hwrm_vnic_qcfg(bp, &vnic, bp->pf.first_vf_id + vf);
+		if (rc)
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt_hwrm.c:1046]: It looks that bd_stall flag can be retrieved correctly here.

File: drivers/net/bnxt/bnxt_ethdev.c (hunk 1 above)
Comment (ajitkhaparde) [reply drivers/net/bnxt/bnxt_ethdev.c]: I suggest moving all these rte_pmd_bnxt_xyz() implementions into its own rte_pmd_bnxt.c 
Are you talking of supporting in bnxt_hwrm.c or HWRM API?
I think HWRM API already allows toggling NO-VEB, VEB, VEPA anyway.
hwrm_func_cfg

Am I missing something?

File: drivers/net/bnxt/bnxt_ethdev.c [hunk 4]
```diff
@@ -819,6 +820,78 @@ static int bnxt_rss_hash_update_op(struct rte_eth_dev *eth_dev,
...
+		rc = bnxt_hwrm_vnic_cfg(bp, &bp->vnic_info[i]);
+		if (rc) {
+			RTE_LOG(ERR, PMD, "Failed to update PF VNIC %d.\n", i);
+			return rc;
+		}
+	}
+
+	/* Stall all active VFs */
+	for (i = 0; i < bp->pf.active_vfs; i++) {
+		rc = bnxt_hwrm_func_vf_stall(bp, i, on);
```
Comment (stephen-hurd) [inline drivers/net/bnxt/bnxt_ethdev.c]: This bit may actually mean the opposite of what on means.  We should check with Hemal about this.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 2 above)
Comment (stephen-hurd) [reply drivers/net/bnxt/bnxt_hwrm.c:923]: The only time traffic should stall is when a ring is full and the hardware mbufs are also full.  You would need to create a test that sets the stall bit then doesn't service the RX ring.  When that happens, other RX rings would stop receiving traffic.

File: drivers/net/bnxt/hsi_struct_def_dpdk.h [hunk 5]
```diff
@@ -4975,6 +5051,141 @@ struct hwrm_vnic_cfg_output {
...
+     */
+    uint64_t resp_addr;
+    /*
+     * This is the host address where the response will be written when the
+     * request is complete. This area must be 16B aligned and must be
+     * cleared to zero before the request is made.
+     */
+    uint32_t enables;
+    /* This bit must be '1' for the vf_id_valid field to be configured. */
+    #define HWRM_VNIC_QCFG_INPUT_ENABLES_VF_ID_VALID           0x1UL
```
Comment (stephen-hurd) [inline drivers/net/bnxt/hsi_struct_def_dpdk.h]: This should be switched to (UINT32_C(0x1)) to be the same as other constants and the correct size.

File: drivers/net/bnxt/hsi_struct_def_dpdk.h [hunk 6]
```diff
@@ -4975,6 +5051,141 @@ struct hwrm_vnic_cfg_output {
...
+     * When a VNIC uses one destination ring group for certain application
+     * (e.g. Receive Flow Steering) where exact match is used to direct
+     * packets to a VNIC with one destination ring group only, there is no
+     * need to configure RSS indirection table for that VNIC as only one
+     * destination ring group is used. When this bit is set to '1', then the
+     * VNIC is enabled in a mode where RSS is enabled in the VNIC using a
+     * RSS context for computing RSS hash but the RSS indirection table is
+     * not configured.
+     */
+    #define HWRM_VNIC_QCFG_OUTPUT_FLAGS_RSS_DFLT_CR_MODE       0x20UL
```
Comment (stephen-hurd) [inline drivers/net/bnxt/hsi_struct_def_dpdk.h]: These constants should use UINT32_C() as well.

**Review by stephen-hurd — CHANGES_REQUESTED** [review]
Looks good for the most part.  I do like the additional vnic fields.

File: drivers/net/bnxt/hsi_struct_def_dpdk.h (hunk 5 above)
Comment (ajitkhaparde) [reply drivers/net/bnxt/hsi_struct_def_dpdk.h]: Let me send you the (master) hwrm file which is adapted to the dpdk submission process.

**Review by ajitkhaparde — COMMENTED** [review]
Stephen has requested changes anyway. Not doubling on that again.

File: drivers/net/bnxt/bnxt_ethdev.c (hunk 1 above)
Comment (steevenlee) [reply drivers/net/bnxt/bnxt_ethdev.c]: From hwrm source, it looks that HWRM API just defined it and not getting it implemented yet.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 2 above)
Comment (steevenlee) [reply drivers/net/bnxt/bnxt_hwrm.c:923]: This works for PF by clearing not servicing BOTH dflt_ring_group and enable bit, but for VF, dflt_ring_group will lost after setting 0.

Comment (steevenlee) [reply drivers/net/bnxt/bnxt_hwrm.c:923]: To make things simple, how about requesting hwrm change by adding "enable" flag option to switch bd_stall only.

File: drivers/net/bnxt/hsi_struct_def_dpdk.h (hunk 5 above)
Comment (steevenlee) [reply drivers/net/bnxt/hsi_struct_def_dpdk.h]: Ajit, that's great to know how to generate the file.

File: drivers/net/bnxt/bnxt_hwrm.c (hunk 2 above)
Comment (steevenlee) [reply drivers/net/bnxt/bnxt_hwrm.c:923]: Got a workaround on this, see latest commit

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 7]
```diff
@@ -855,12 +855,14 @@ int bnxt_hwrm_vnic_cfg(struct bnxt *bp, struct bnxt_vnic_info *vnic)
...
-			     HWRM_VNIC_CFG_INPUT_ENABLES_MRU);
+	req.enables = rte_cpu_to_le_32( HWRM_VNIC_CFG_INPUT_ENABLES_RSS_RULE
+			| HWRM_VNIC_CFG_INPUT_ENABLES_MRU);
 	req.vnic_id = rte_cpu_to_le_16(vnic->fw_vnic_id);
-	req.dflt_ring_grp = rte_cpu_to_le_16(vnic->dflt_ring_grp);
+	if (!vnic->bd_stall) {
+		req.enables |=
+				rte_cpu_to_le_32(HWRM_VNIC_CFG_INPUT_ENABLES_DFLT_RING_GRP);
+		req.dflt_ring_grp = rte_cpu_to_le_16(vnic->dflt_ring_grp);
+	}
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt_hwrm.c]: This works well for both PF and VF now

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 8]
```diff
@@ -1975,3 +1976,21 @@ int bnxt_hwrm_allocate_vfs(struct bnxt *bp, int num_vfs)
...
+{
+	struct hwrm_func_cfg_input req = {0};
+	struct hwrm_func_cfg_output *resp = bp->hwrm_cmd_resp_addr;
+	int rc;
+
+	HWRM_PREP(req, FUNC_CFG, -1, resp);
+
+	req.fid = rte_cpu_to_le_16(0xffff);
+	req.enables = rte_cpu_to_le_32(HWRM_FUNC_CFG_INPUT_ENABLES_EVB_MODE);
+	req.evb_mode = bp->pf.evb_mode;
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt_hwrm.c:2003]: I can see VEB mode works, while VEPA does not...

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 9]
```diff
@@ -1857,6 +1857,7 @@ static int update_pf_resource_max(struct bnxt *bp)
 	HWRM_CHECK_RESULT;
 
 	bp->max_tx_rings = rte_le_to_cpu_16(resp->alloc_tx_rings);
+	bp->pf.evb_mode = resp->evb_mode;
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt_hwrm.c:1869]: Always return 0, although it looks that NIC is forwarding as VEB mode.

File: drivers/net/bnxt/bnxt_hwrm.c [hunk 10]
```diff
@@ -1036,6 +1036,10 @@ int bnxt_hwrm_func_vf_stall(struct bnxt *bp, uint16_t vf, uint8_t on)
 		if (rc)
 			break;
 
+		if (vnic.dflt_ring_grp)
+			bp->pf.vnic_dflt_ring_group[vnic.fw_vnic_id] = vnic.dflt_ring_grp;
+		else
+			vnic.dflt_ring_grp = bp->pf.vnic_dflt_ring_group[vnic.fw_vnic_id];
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt_hwrm.c]: Has to set dflt_ring_group to 0 to make bd_stall work, next qcfg will return 0, has to cache here.

File: drivers/net/bnxt/bnxt.h [hunk 11]
```diff
@@ -73,6 +73,7 @@ struct bnxt_pf_info {
 #define BNXT_EVB_MODE_VEB	1
 #define BNXT_EVB_MODE_VEPA	2
 	uint8_t 		evb_mode;
+	uint16_t 		vnic_dflt_ring_group[340];
```
Comment (steevenlee) [inline drivers/net/bnxt/bnxt.h:82]: It looks that ring group range from 170 to 340, not sure how to calculate the start and stop.

**Review by stephen-hurd — APPROVED** [review]

**Pull request merged**